src/main/java/com/yahoo/bullet/bql/query/LayeredSchema.java

```diff
@@ -15,82 +15,201 @@
+     * on top, their depths will not be adjusted. It is recommended to add layers from the top.
+     *
+     * @param newSchema The new {@link Schema} to add to the top layer.
+     * @param newAliases The new {@link Map} of aliases to add to the top layer.
+     */
     public void addLayer(Schema newSchema, Map<String, String> newAliases) {
-        subSchema = new LayeredSchema(schema, aliases, subSchema, locked);
+        subSchema = new LayeredSchema(schema, aliases, subSchema, depth, locked);
         schema = newSchema;
         aliases = newAliases;
         locked = false;
+        subSchema.increaseDepth();
     }
 
+    /**
+     * Locks this {@link LayeredSchema}, preventing access to all layers below.
+     */
     public void lock() {
         locked = true;
     }
 
-    public Schema.Field getField(String field) {
-        if (schema == null) {
-            return null;
-        }
-        Type type = schema.getType(field);
-        if (type != Type.NULL) {
-            return new Schema.PlainField(field, type);
-        }
-        String alias = aliases.get(field);
-        if (alias != null) {
-            return new Schema.PlainField(alias, schema.getType(alias));
-        }
-        return subSchema != null && !subSchema.locked ? subSchema.getField(field) : null;
+    /**
+     * Unlocks this {@link LayeredSchema}, allowing access to layers below.
+     */
+    public void unlock() {
+        locked = false;
     }
 
-    public Type getType(String field) {
+    /**
+     * Gets the current depth of this {@link LayeredSchema}. Depth is defined starting at 0 for the top layer and
+     * increases as you go deeper.
+     *
+     * @return The depth of this layer.
+     */
+    public int depth() {
+        return depth;
+    }
+
+    /**
+     * Searches for the given field from this layer. The minimum depth parameter can be provided to ensure that
+     * the field, if found, is at that depth or greater. The depth is the depth of this layered schema as defined by
+     * {@link #depth()}. This can be used to skip layers for the search.
+     *
+     * @param field The field to search for.
+     * @param minimumDepth The minimum (whole number) for the depth to find the field from.
+     * @return A {@link FieldLocation} for the field. It is non-null. If the schema does not exist, the type will be
+     *         {@link Type#UNKNOWN}. If field is not found, the type will be be {@link Type#NULL}.
+     */
+    public FieldLocation findField(String field, int minimumDepth) {
         if (schema == null) {
             // If the schema is null, ignore the subschema and just return Type.UNKNOWN
-            return Type.UNKNOWN;
+            return FieldLocation.from(null, Type.UNKNOWN, depth);
         }
         Type type = schema.getType(field);
-        if (type != Type.NULL) {
-            return type;
+        if (type != Type.NULL && depth >= minimumDepth) {
```

Comment:
Could we just skip to the next layer if depth < minimumDepth?

Reply:
Sure pulled out the depth check